### PR TITLE
python fastpath for DTensor detach(), confirm that aliasing DTensorSpec is ok

### PR DESCRIPTION
**Pull request by bdhirsh**
My goal right now is to try to make the "vanilla" AccumulateGrad path for DTensor (that just calls detach) fast. I'm doing this in two steps:

(1) [this PR]: hardcode aten.detach in DTensor to re-use the input tensor's DTensorSpec, instead of running "real" sharding prop.

(2) [assuming success of 1]: move the detach() call into C++, try adding a DTensor dispatch key, and avoid dispatching back to python entirely (except for some code that probably needs to allocate a pyobject for the output DTensor, from C++)

I'm pushing this PR first to confirm that I don't break anything with my detach fastpath. I did some manual local testing to confirm that for normal usages of detach, the input and output DTensor have equal DTensorSpec objects. Technically, we previously would allocate a fresh DTensorSpec, and with this change we are just re-using the input tensor's DTensorSpec. So I'm mostly hoping that DTensorSpecs don't generally get mutated

This by itself does seem to speed up `alias` by quite a bit (roughly 2.5x speedup, from ~336us -> 133us):

**aten.detach(plain_tensor)**
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7f8da2921790>
_ = x.detach()
  4.80 us
  1 measurement, 100000 runs , 1 thread
```

**aten.detach(DTensor) [before this PR]**
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7f47cd68e750>
_ = x_dt.detach()
  336.40 us
  1 measurement, 1000 runs , 1 thread
```

**aten.detach(DTensor) [after this PR]**
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7f0a34c05520>
_ = x_dt.detach()
  Median: 133.45 us
  2 measurements, 1000 runs per measurement, 1 thread
```

benchmark script:
```
import torch
import torch.distributed as dist
from torch.distributed.tensor import DeviceMesh, DTensor, Partial, Replicate, Shard
from torch.testing._internal.distributed.fake_pg import FakeStore
import torch.utils.benchmark as benchmark

fake_store = FakeStore()
dist.init_process_group("fake", store=fake_store, rank=0, world_size=2)

mesh = torch.distributed.device_mesh.init_device_mesh('cuda', (2,))
x = torch.randn(4, 4, requires_grad=True)
x_dt = DTensor.from_local(x, mesh, [Shard(0)], run_check=False)

t0 = benchmark.Timer(
    stmt='_ = x_dt.detach()',
    globals={'x_dt': x_dt},
)
print(t0.blocked_autorange())


dist.destroy_process_group()
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #160580

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

### CODE REVIEW

**Comment by wconstab** [comment]
How are you thinking about landing something like this?

It seems like this change is self-contained enough and useful enough to merit landing as-is, unless there is an even faster version of it in c++ coming imminently.

Otoh, it also seems like there could be lots more cases we'd want to optimize, and that would probably lead to having a more extensible framework for opting more ops into this path.

Also, it'd be good to have some way of measuring overhead in CI but that's a side point.  Pei worked on some actual benchmarking.  But nobody has tried instruction counting yet, and that might actually be viable here (haven't checked how expensive it is to do that)

**Comment by bdhirsh** [comment]
I would be ok with landing it as-is (maybe even going back to just the detach() case if that is the main source of slowness we have to worry about r.e. AccumulateGrad slowness).

I'm hoping that this is "enough" and we won't have to reach into making C++ changes, but in either case, it would be easy enough to land this and remove it later.

> Otoh, it also seems like there could be lots more cases we'd want to optimize, and that would probably lead to having a more extensible framework for opting more ops into this path.

Fair, if we want to add a lot of fastpaths in the future this may get unwieldy pretty quickly. I am sort of hoping that we can draw the line as "AccumulateGrad is tough to compile so lets try to make DTensor overheads for it fast, and compile everything else". But I'm not sure how likely it is that we can do that.

**Comment by ezyang** [comment]
This is still failing lots of tests. What's the plan?

**Comment by bdhirsh** [comment]
Spot checked a few of the tests locally and they also fail when I remove the fastpath - I'll rebase and check for new failures

File: torch/distributed/tensor/_api.py [hunk 1]
```diff
@@ -346,6 +356,9 @@ def __coerce_same_metadata_as_tangent__(self, flatten_spec, expected_type=None):
     # pyre-fixme[3]: Return type must be annotated.
     # pyre-fixme[2]: Parameter must be annotated.
     def __torch_dispatch__(cls, func, types, args=(), kwargs=None):  # type: ignore[override]
+        fastpath_out = _maybe_fastpath_torch_dispatch(func, args, kwargs or {})
```
Comment (ezyang) [inline torch/distributed/tensor/_api.py]: Instead of doing a function call, you might consider just doing the condition inline and save a few more time

**Review by ezyang — APPROVED** [review]
This is kind of risky so monitor on rollout, I think

**Comment by swolchok** [comment]
are we expecting this to re-land?

**Comment by ezyang** [comment]
Not until Brian is back from PTO (2 weeks) at least
